### ISSUE TITLE
DnD5e - Need a way for players to denote advantage on initiative

### ISSUE THREAD

**Issue by trey-kirk-sp** [mr-ice/maptool-macros#248]
The gap is that DMs roll initiative for all tokens and some tokens have advantage. The Roll Initiative macros have no concept of determining that, and I think it should stay that way. I think we need to create a new Expression type "Initiative" and designate a State to indicate Advantage on Initiative. 

- [x] Designate an existing state to rename to ”Advantage to Initiative"
- [x] Create a new RE prototype with Expression type "Initiative". It's a 1d20 roll with typ Types Advantagable and Basic.
- [x] Refactor Initiative rolling macros to use the RE and DR
- [x] Extend DiceRoller_applyConditions to connect this state to having advantage when rolled. 
- [x] See if it makes sense to update the Set Conditions macro to include this State.
- [x] This is the first of what could be a myriad of situations where an Expression type has adv/dis. Work up a theoretical scenario of 10 things like this and see if it becomes an unwieldy framework.

**Comment by trey-kirk-sp** [comment]
Some thoughts I had on the run and don't wanna forget:
- refactor DR_applyConditions to use an open ended framework for determining appliers
- an applier is a function using a naming convention: somePackage_RollExpression_ThingClassification_Thing_apply. As an example, for this idea, if the State we're called "InitiativeAdvantage" then it will be something like:
‘dnd5e_RollExpression_State_InitiativeAdvantage_apply‘

This sets us up to add drop-in resources later that might define how to apply class features to a given roll. Those resources' apply functions can not only read States, but any information on the token to determine how to modify the RE. In a future iteration, configuration of these abstract elements could be addressed with the Character Sheet. Drop-ins would need to bring a renderer of Features for the char sheet and subprocessors to consume those elements. With that, we could cover actions, such as a player marks their character sheet as enabling some ability to grant advantage on Ability checks and their Ability checks automatically pick up the advantage.

**Comment by havoclad** [comment]
<img width="850" alt="Screen Shot 2021-03-20 at 11 27 18 AM" src="https://user-images.githubusercontent.com/9877645/111877006-3f38e300-896f-11eb-8b62-ed05a77c92d0.png">

I think this feature of the Twilight Domain is the most annoying that I've seen for this feature.  Players need to be able to set their advantage manually, and bonus points if it unapplies after an initiative roll. :)

**Comment by trey-kirk-sp** [comment]
Think I've taken this as far as I should, at least for the time being. I've created a slew of States which mark advantage for a slew of actions (ability checks, initiative, attacks). However, application of those states would have to be initiated by the token owner. Meaning, if I did write a macro for Vigilant Blessing, the casting player would only be able to apply "Advantage on Initiative" to their own token. It would not be able to apply it to any other token without the user having ownership of those tokens. To circumvent that, I'd have to write a trusted macro that's part of the 5e library, and that feels like an overly specialized macro that doesn't have any API usability or general application to all tokens. Perhaps down the road, a relevant Class specific library would include it.

However, if an industrious user wanted to write a personal macro to enable/disable this state:
```
<!-- Enables the state -->
[h: setState ("Advantage on Initiative", 1)]
<!-- Disables the state -->
[h: setState ("Advantage on Initiative", 0)]
```

Breaking off the UI component of this to another issue. For now, users apply these states via the States menu on the token. It's a bit busy, but I've at least organized it in a manner that makes the desired state easier to find.

Attack related states work fine. However, only attacks from Attack Editor take full advantage of them. O5E Monsters need to have their attacks updated to use the Weapon type. In the same vein, AE2 doesn't build Weapon type attacks, either. I'll update the O5e parser in another issue. Jay will need to tackle AE2, if he wants to.
